### PR TITLE
Upgrade to swift 4

### PR DESCRIPTION
**Pull request by bsudekum**
This begins the process of upgrading to swift 4. 

/cc @1ec5 @ericrwolfe @frederoni

### CODE REVIEW

**Comment by bsudekum** [comment]
> enclosing any incompatible code in #if swift checks?

Let's see how many there are. This could become very tiresome if it's a lot of code. The other option is to maintain a branch for x more months... and with a set EOL.

File: MapboxCoreNavigationTests/GeometryTests.swift [hunk 1]
```diff
@@ -190,7 +190,7 @@ class GeometryTests: XCTestCase {
         
         // https://github.com/Turfjs/turf/blob/142e137ce0c758e2825a260ab32b24db0aa19439/packages/turf-distance/test.js
         let a = distance(along: line)
-        XCTAssertEqualWithAccuracy(a, 97_159.57803131901, accuracy: 1)
+        XCTAssertEqual(a, 97_159.57803131901, accuracy: 1)
```
Comment (frederoni) [inline MapboxCoreNavigationTests/GeometryTests.swift]: This won't build using an earlier version.
Try adding: https://github.com/mapbox/turf-swift/blob/e17d1d05c08796536dea0b36a38b9c2bbde0fb06/TurfTests/TurfTests.swift#L7-L12

**Comment by frederoni** [comment]
Should we add an additional Bitrise app with the new Xcode 9 stack?

**Comment by bsudekum** [comment]
Yes, we should test both.

**Comment by bsudekum** [comment]
Added bitrise test for xcode 9. It will only run on this branch.

**Comment by bsudekum** [comment]
So I added quite a few `@objc` declarations but there are still a handful that are not bridging.

**Comment by bsudekum** [comment]
A line line https://github.com/mapbox/mapbox-navigation-ios/pull/663/files#diff-851f438872328e98cbe395ac0733c16bR130 will not accept the prefix `@objc` because it doesn't bridge. However, I guess it never bridged to begin with?

**Comment by 1ec5** [comment]
Not sure which line you’re referring to, but there are probably a number of methods that never bridged to Objective-C. Let’s track those issues separately. Making these situations explicit is one of the benefits of Swift 4’s `@objc` behavior.

File: MapboxCoreNavigation/NavigationLocationManager.swift [hunk 2]
```diff
@@ -17,13 +17,13 @@ open class NavigationLocationManager: CLLocationManager {
     /**
      Indicates whether the device is plugged in or not.
      */
-    public private(set) var isPluggedIn: Bool = false
+    @objc public private(set) var isPluggedIn: Bool = false
```
Comment (JThramer) [inline MapboxCoreNavigation/NavigationLocationManager.swift]: >`private(set)`

I did not know you could do that.

**Comment by bsudekum** [comment]
We'll need to update MapboxDirections.swift to swift 4 before merging this.

**Comment by bsudekum** [comment]
This is blocked by https://github.com/CocoaPods/CocoaPods/issues/7134

**Comment by 1ec5** [comment]
Is it really blocked? This PR was designed to work when compiled as Swift 3, so we could have Carthage compiling it as Swift 4 and CocoaPods compiling it as Swift 3 for the time being.

**Comment by bsudekum** [comment]
@1ec5 I guess that would work.

File: Cartfile [hunk 3]
```diff
@@ -2,7 +2,7 @@ binary "https://www.mapbox.com/ios-sdk/Mapbox-iOS-SDK.json" ~> 3.6
 github "mapbox/MapboxDirections.swift" ~> 0.12
 github "mapbox/turf-swift" ~> 0.0.3
 github "rs/SDWebImage" ~> 4.1
-github "Project-OSRM/osrm-text-instructions.swift" ~> 0.5
+github "Project-OSRM/osrm-text-instructions.swift" "f60d5bc175794aee6c7199d60548820e0fb23097"
```
Comment (bsudekum) [inline Cartfile]: Will change when https://github.com/Project-OSRM/osrm-text-instructions.swift/pull/50 is released.

**Comment by bsudekum** [comment]
@mapbox/navigation-ios this is ready for review.

File: Examples/Objective-C/ViewController.m [hunk 4]
```diff
@@ -126,7 +126,6 @@ -(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
         
         MGLPointAnnotation *destination = [[MGLPointAnnotation alloc] init];
         destination.coordinate = self.destination;
-        controller.destination = destination;
```
Comment (1ec5) [inline Examples/Objective-C/ViewController.m]: Why was this line removed? Doesn’t this cause the preceding two lines to be useless?

File: MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift [hunk 5]
```diff
@@ -21,7 +21,7 @@ class MapboxCoreNavigationTests: XCTestCase {
         navigation.resume()
         let depart = CLLocation(coordinate: CLLocationCoordinate2D(latitude: 37.795042, longitude: -122.413165), altitude: 1, horizontalAccuracy: 1, verticalAccuracy: 1, course: 0, speed: 10, timestamp: Date())
         
-        self.expectation(forNotification: RouteControllerDidPassSpokenInstructionPoint.rawValue, object: navigation) { (notification) -> Bool in
+        self.expectation(forNotification: NSNotification.Name(rawValue: RouteControllerDidPassSpokenInstructionPoint.rawValue), object: navigation) { (notification) -> Bool in
```
Comment (1ec5) [inline MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift]: Use `RouteControllerDidPassSpokenInstructionPoint` directly instead of converting to a raw value and back to an `NSNotification.Name` (×4).

File: MapboxNavigation/BottomBannerViewLayout.swift [hunk 6]
```diff
@@ -6,13 +6,13 @@ extension BottomBannerView {
         
         let timeRemainingLabel = TimeRemainingLabel()
         timeRemainingLabel.translatesAutoresizingMaskIntoConstraints = false
-        timeRemainingLabel.font = .systemFont(ofSize: 28, weight: UIFontWeightMedium)
+        timeRemainingLabel.font = .systemFont(ofSize: 28, weight: UIFont.Weight.medium)
```
Comment (1ec5) [inline MapboxNavigation/BottomBannerViewLayout.swift]: Nit: drop `UIFont.Weight` (×2).

File: MapboxNavigation/DayStyle.swift [hunk 7]
```diff
@@ -58,22 +58,22 @@ open class DayStyle: Style {
             tintColor = .defaultTint
         }
         
-        ArrivalTimeLabel.appearance().font = UIFont.systemFont(ofSize: 18, weight: UIFontWeightMedium).adjustedFont
+        ArrivalTimeLabel.appearance().font = UIFont.systemFont(ofSize: 18, weight: UIFont.Weight.medium).adjustedFont
```
Comment (1ec5) [inline MapboxNavigation/DayStyle.swift]: Nit: drop `UIFont.Weight` (×10).

File: MapboxNavigation/InstructionLabel.swift [hunk 8]
```diff
@@ -57,8 +57,8 @@ open class InstructionLabel: StylableLabel {
         attributedText = string
     }
     
-    var attributes: [String: Any] {
-        return [NSFontAttributeName: font, NSForegroundColorAttributeName: textColor]
+    var attributes: [NSAttributedStringKey: Any] {
+        return [NSAttributedStringKey.font: font, NSAttributedStringKey.foregroundColor: textColor]
```
Comment (1ec5) [inline MapboxNavigation/InstructionLabel.swift]: Nit: drop `NSAttributedStringKey`.

File: MapboxNavigation/InstructionsBannerViewLayout.swift [hunk 9]
```diff
@@ -119,13 +119,13 @@ extension BaseInstructionsBannerView {
     func setupAvailableBounds() {
         // Abbreviate if the instructions do not fit on one line
         primaryLabel.availableBounds = {
-            let height = ("|" as NSString).size(attributes: [NSFontAttributeName: self.primaryLabel.font]).height
+            let height = ("|" as NSString).size(withAttributes: [NSAttributedStringKey.font: self.primaryLabel.font]).height
```
Comment (1ec5) [inline MapboxNavigation/InstructionsBannerViewLayout.swift]: Nit: drop `NSAttributedStringKey` (×2).

File: MapboxNavigation/NextBannerView.swift [hunk 10]
```diff
@@ -40,7 +40,7 @@ open class NextBannerView: UIView {
         self.instructionLabel = instructionLabel
         
         instructionLabel.availableBounds = {
-            let height = ("|" as NSString).size(attributes: [NSFontAttributeName: self.instructionLabel.font]).height
+            let height = ("|" as NSString).size(withAttributes: [NSAttributedStringKey.font: self.instructionLabel.font]).height
```
Comment (1ec5) [inline MapboxNavigation/NextBannerView.swift]: Nit: drop `NSAttributedStringKey`.

File: MapboxNavigation/Style.swift [hunk 11]
```diff
@@ -193,18 +193,18 @@ open class StylableLabel : UILabel {
...
     {
         didSet { update() }
     }
-    dynamic public var unitTextColor: UIColor = #colorLiteral(red: 0.6274509804, green: 0.6274509804, blue: 0.6274509804, alpha: 1)
+    @objc dynamic public var unitTextColor: UIColor = #colorLiteral(red: 0.6274509804, green: 0.6274509804, blue: 0.6274509804, alpha: 1)
     {
         didSet { update() }
     }
-    dynamic public var valueFont: UIFont = UIFont.systemFont(ofSize: 16, weight: UIFontWeightMedium) {
+    @objc dynamic public var valueFont: UIFont = UIFont.systemFont(ofSize: 16, weight: UIFont.Weight.medium) {
```
Comment (1ec5) [inline MapboxNavigation/Style.swift]: Nit: drop `UIFont.Weight` (×3).

File: MapboxNavigation/Style.swift [hunk 12]
```diff
@@ -231,11 +231,11 @@ open class DistanceLabel: StylableLabel {
             return
         }
 
-        let valueAttributes: [String: Any] = [NSForegroundColorAttributeName: valueTextColor, NSFontAttributeName: valueFont]
-        let unitAttributes: [String: Any] = [NSForegroundColorAttributeName: unitTextColor, NSFontAttributeName: unitFont]
+        let valueAttributes: [NSAttributedStringKey: Any] = [NSAttributedStringKey.foregroundColor: valueTextColor, NSAttributedStringKey.font: valueFont]
+        let unitAttributes: [NSAttributedStringKey: Any] = [NSAttributedStringKey.foregroundColor: unitTextColor, NSAttributedStringKey.font: unitFont]
```
Comment (1ec5) [inline MapboxNavigation/Style.swift]: Nit: drop `NSAttributedStringKey` from both these lines.

File: MapboxNavigation/UIFont.swift [hunk 13]
```diff
@@ -37,14 +37,14 @@ extension UIFont {
     
     func with(fontFamily: String?) -> UIFont {
         guard let fontFamily = fontFamily else { return self }
-        let weight = (fontDescriptor.object(forKey: UIFontDescriptorTraitsAttribute) as! [String:Any])[UIFontWeightTrait]
-        let attributes = [UIFontDescriptorTraitsAttribute: [UIFontWeightTrait: weight]]
+        let weight = (fontDescriptor.object(forKey: UIFontDescriptor.AttributeName.traits) as! [String:Any])[UIFontDescriptor.TraitKey.weight.rawValue]
```
Comment (1ec5) [inline MapboxNavigation/UIFont.swift]: Nit: drop `UIFontDescriptor.AttributeName`.

File: MapboxNavigation/UIFont.swift [hunk 14]
```diff
@@ -37,14 +37,14 @@ extension UIFont {
...
+        let attributes = [UIFontDescriptor.AttributeName.traits: [UIFontDescriptor.TraitKey.weight: weight]]
```
Comment (1ec5) [inline MapboxNavigation/UIFont.swift]: Inline this variable so that you can drop `UIFontDescriptor.AttributeName` and possibly `UIFontDescriptor.TraitKey`.

File: MapboxNavigation/UIFont.swift [hunk 15]
```diff
@@ -37,14 +37,14 @@ extension UIFont {
...
         let descriptor = UIFontDescriptor(name: fontName, size: pointSize).withFamily(fontFamily).addingAttributes(attributes)
         return UIFont(descriptor: descriptor, size: pointSize)
     }
     
     func with(weight: CGFloat) -> UIFont {
-        let font = UIFont(descriptor: fontDescriptor.addingAttributes([UIFontWeightTrait: weight]), size: pointSize)
+        let font = UIFont(descriptor: fontDescriptor.addingAttributes([UIFontDescriptor.AttributeName.traits: weight]), size: pointSize)
```
Comment (1ec5) [inline MapboxNavigation/UIFont.swift]: Nit: drop `UIFontDescriptor.AttributeName`.

File: MapboxNavigationTests/InstructionsBannerViewTests.swift [hunk 16]
```diff
@@ -131,7 +131,7 @@ extension InstructionsBannerViewTests {
         
         view.distanceLabel.valueFont = UIFont.systemFont(ofSize: 24)
         view.distanceLabel.unitFont = UIFont.systemFont(ofSize: 14)
-        view.primaryLabel.font = UIFont.systemFont(ofSize: 30, weight: UIFontWeightMedium)
-        view.secondaryLabel.font = UIFont.systemFont(ofSize: 26, weight: UIFontWeightMedium)
+        view.primaryLabel.font = UIFont.systemFont(ofSize: 30, weight: UIFont.Weight.medium)
+        view.secondaryLabel.font = UIFont.systemFont(ofSize: 26, weight: UIFont.Weight.medium)
```
Comment (1ec5) [inline MapboxNavigationTests/InstructionsBannerViewTests.swift]: Nit: drop `UIFont.Weight` from both these lines.

File: Examples/Objective-C/ViewController.m (hunk 4 above)
Comment (bsudekum) [reply Examples/Objective-C/ViewController.m]: `destination` was deprecated eons ago.

**Comment by bsudekum** [comment]
@1ec5 fixed.

File: MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift [hunk 17]
```diff
@@ -43,7 +43,7 @@ class MapboxCoreNavigationTests: XCTestCase {
         let locationManager = ReplayLocationManager(locations: [location, location])
         let navigation = RouteController(along: route, directions: directions, locationManager: locationManager)
         
-        self.expectation(forNotification: NSNotification.Name(rawValue: RouteControllerDidPassSpokenInstructionPoint.rawValue), object: navigation) { (notification) -> Bool in
+        self.expectation(forNotification: RouteControllerDidPassSpokenInstructionPoint), object: navigation) { (notification) -> Bool in
```
Comment (1ec5) [inline MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift]: Unbalanced parentheses are causing a build failure (×3).

File: Examples/Objective-C/ViewController.m (hunk 4 above)
Comment (1ec5) [reply Examples/Objective-C/ViewController.m]: So shouldn’t the two lines above be removed too?

**Comment by 1ec5** [comment]
Fixed a serving or two of copypasta in fcb0e12c0434a574d77a655c25723004c7b3f27b and ac5f85e2e8da6c710a0f902cedecacd4f83a6eb6.
